### ISSUE TITLE
Authentication documentation:  `authCallback` and Auth Token tutorial for javascript

### ISSUE THREAD

**Issue by ben-xD** [ably/docs#1094]
I couldn’t find any documentation or example code on using `authCallback` properly (existing solutions use authUrl), but finally figured it out (after juggling it over 2 days while battling with other issues cors, webworkers/ cloudflare workers), it turns out to be quite simple, though there are _many ways to get it wrong_, leading to unhelpful error messages. I found that the docs explain the high level overview, and the first part of the code required, but I wish there were more code examples. 

I think it would be nice to add something like this (though you may choose to improve it) to the auth docs ([2.1 Ably TokenRequest is created by your servers and passed to clients](https://ably.com/documentation/core-features/authentication#token-request-process)). The code lives in this file https://github.com/ben-xD/Club/blob/fdf5dd336b197ac43c70c87cb58e7afc49f3f7d2/web/public/ts/Messaging.ts#L52
```js
            this.ablyClient = new Realtime({
                authCallback: async (data, callback) => {
                    try {
                        const tokenRequest = // Get your token request from your backend via a http request
                        callback(null, tokenRequest)
                    } catch (e) {
                        callback(e, null)
                    }
                }
            })
```

TLDR:
- I propose adding more detail to using `authCallback` docs to https://ably.com/documentation/core-features/authentication#token-request-process
- Add authCallback as an option for javascript clients under each programming language that has js clients: https://ably.com/tutorials/token-authentication
- Extra suggestion: I found references to testing-only code quite confusing": e.g. `The last of those (providing a literal token or tokenDetails) is mostly only used for testing`.  and just having the happy production path documentation working perfectly would be great.

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/EDU-1798) by [Unito](https://www.unito.io)

**Comment by lmars** [comment]
Just to add an extra data point, an example of using `authCallback` can be found in the [Serverless Auth Example](https://github.com/ably-labs/ably-serverless-auth) I wrote recently, specifically in [lambda/web/app.js](https://github.com/ably-labs/ably-serverless-auth/blob/main/lambda/web/app.js#L93-L134)

**Comment by paddybyers** [comment]
Can you also way what the unhelpful error messages were?

**Comment by tbedford** [comment]
@ben-xD Thanks for this. I also did some initial playing around with auth callback [here](https://github.com/tbedford/ably-time-server/blob/main/client/iot-callback.js). We haven't had time to write this up yet, but it's on the todo list. Thanks for raising!

**Comment by ben-xD** [comment]
> Can you also say what the unhelpful error messages were?

Yep,  here are a few issues I've just reproduced:
- When I added a `TokenRequest` (as according to the [docs](https://ably.com/documentation/core-features/authentication#token-authentication)) in the `AuthOptions`, I got an error message about token being undefined. Now it turns out me **overriding** the typescript warnings for that was wrong, because JS simply didn't implement passing TokenRequest as a token: https://github.com/ably/ably-js/issues/607 It would have been good to be warned that TokenRequest is not implemented in ably-js. I was confused because I thought constructing Ably.Rest or Ably.Realtime requires authentication (a token) passed as an AuthOption, even if authUrl and authCallback was being used. To call the `this.ablyRealtimeClient.auth.requestToken`, you already need the client created/ connected. It turns out, using _just_ authCallback is enough.
![Frame 97](https://user-images.githubusercontent.com/24711048/118469116-53922580-b6fd-11eb-9e4d-16cf14697895.png)

- In another case, when I passed `authCallback` in the options, I got an error, roughly saying: key, authUrl or authCallback was not defined. I can't reproduce this now though.

**Comment by rtpa25** [comment]
```
Frontend NEXT
"user client";

import Ably from "ably";
import { AblyProvider } from "ably/react";
import * as React from "react";

import { api } from "~/trpc/server";

const ablyClient = new Ably.Realtime({
  authCallback: (_tokenParams, callback) => {
    api.socket.genToken
      .mutate()
      .then(({ tokenRequest }) => {
        callback(null, tokenRequest);
      })
      .catch((err) => {
        callback(err, null);
      });
  },
  useTokenAuth: true,
  echoMessages: false,
  closeOnUnload: false,
});

export const AblyContextProvider = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  return <AblyProvider client={ablyClient}>{children}</AblyProvider>;
};
```

```
Backend tRPC on Fastify

import { createTRPCRouter, privateProcedure } from "../trpc";

export const socketRouter = createTRPCRouter({
// privateProcedure checks request cookies and decides if the user is authed
  genToken: privateProcedure.mutation(async ({ ctx }) => {
   // I have already instantiated ably somewhere else 
    const tokenReq = await ably.auth.createTokenRequest({
      clientId: ctx.user.id,
    });
    return {
      tokenRequest: tokenReq,
    };
  }),
});
 
```


With this setup according to the docs the initial call made by my client passes and ably works according to the network tab
![Screenshot 2024-05-16 at 4 11 35 PM](https://github.com/ably/docs/assets/72537293/c00f2552-c8b1-4bf5-9ca4-0f41ae417204)
while it keeps on making subsequent requests from the sdk without the cookie hence auth fails how does one setup such that it sends the cookie
<img width="1380" alt="Screenshot 2024-05-16 at 4 12 26 PM" src="https://github.com/ably/docs/assets/72537293/5881e573-3246-49d3-8721-d23f11ec36d8">
